feat: GTFS related endpoints

jujiang526 opened the pull request:
**Summary:**
[Linked Issue](https://github.com/MobilityData/mobility-feed-api/issues/33)

Implemented the following endpoints without supporting filtering or sorting

[Get GTFS feeds](https://mobilitydata.github.io/mobility-feed-api/SwaggerUI/index.html#/feeds/getGtfsFeeds)
[Get GTFS RT feeds](https://mobilitydata.github.io/mobility-feed-api/SwaggerUI/index.html#/feeds/getGtfsRtFeeds)
[Get GTFS feed by ID](https://mobilitydata.github.io/mobility-feed-api/SwaggerUI/index.html#/feeds/getGtfsFeed)
[Get GTFS RT feed by ID](https://mobilitydata.github.io/mobility-feed-api/SwaggerUI/index.html#/feeds/getGtfsRtFeed)
[Get GTFS feed datasets for a given feed ID](https://mobilitydata.github.io/mobility-feed-api/SwaggerUI/index.html#/feeds/getGtfsFeedDatasets)
[Get GTFS datasets by dataset ID](https://mobilitydata.github.io/mobility-feed-api/SwaggerUI/index.html#/datasets/getDatasetGtfs)

**Expected behavior:** 

Tested all those end points. The performance is good. Added a few unit tests regarding bounding filtering and database join query.


Please make sure these boxes are checked before submitting your pull request - thanks!

- [X] Run the unit tests with `./scripts/api-tests.sh` to make sure you didn't break anything
- [X] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [X] Linked all relevant issues
- [X] Include screenshot(s) showing how this pull request works and fixes the issue(s)

aronza commented:
Hi @jujiang526 ,

Thank you for all this! I've refactored it further to reduce code duplication.

Also I saw that you were mapping database feed or dataset IDs directly to API IDs. I believe @jcpitre clarified in an earlier meeting that they want us to use stable_id when accepting or returning IDs in API. I've made the changes in the code and verified in my local Postman it works as expected.

Though the tests you've added were using these database IDs that are auto-generated in https://github.com/MobilityData/mobility-feed-api/blob/02f2b6f0b2a4dddac51f7e02de220c965e9004f4/api/src/scripts/populate_db.py#L123C1-L123C1 in our locals, so these tests won't pass anywhere else. Do you know which dataset stable ID did you use for these values?

indraneel commented:
🚨 Do not merge this PR into main, **until** #94 is reviewed/merged into this branch

aronza commented:
> 🚨 Do not merge this PR into main, **until** #94 is reviewed/merged into this branch

merged and resolved the conflicts from main

jcpitre commented:
Added https://github.com/MobilityData/mobility-feed-api/issues/96 to track filtering on dates for feeds.

jcpitre commented on api/src/database/database.py:
```diff
@@ -120,7 +128,10 @@ def select(self, model: Type[Base] = None, query: Select = None,
-            return self.session.execute(query).all()
+            results = self.session.execute(query).all()
+            if group_by:
+                return [list(group) for _, group in itertools.groupby(results, group_by)]
```
If I am not mistaken, groupby needs a list sorted by the key (in our case stable_id).
I just want to make sure it's sorted. 
If it is, I suggest you put a comment like: "Results are sorted by stable_id since we used the ORDER BY in such and such place", or whatever mean was used to sort.

indraneel replied:
@aronza to take this

aronza replied:
Thanks for pointing this out @jcpitre , I've added order by clause to queries using groupby

jcpitre commented on docs/DatabaseCatalogAPI.yaml:
```diff
@@ -78,7 +80,10 @@ paths:
+        - $ref: "#/components/parameters/country_code"
+        - $ref: "#/components/parameters/subdivision_name"
+        - $ref: "#/components/parameters/municipality"
         - $ref: "#/components/parameters/sort"
```
There's also filtering by producer_URL according to https://github.com/MobilityData/mobility-feed-api/issues/33
It could be delegated to a new issue.

aronza replied:
added to all feeds endpoints

jujiang526 commented:
@aronza - Can you please try to run the unit test in your environment? I believe I have fixed it. I also fixed a small issue with returning duplicated records from the joined table.

aronza commented:
> @aronza - Can you please try to run the unit test in your environment? I believe I have fixed it. I also fixed a small issue with returning duplicated records from the joined table.

They pass on my local too @jujiang526 . Though our changes to database class broke @goorui 's unit tests, so I refactored these tests to work with the latest changes

aronza commented on api/src/database/database.py:
```diff
@@ -24,6 +24,12 @@ class Database:
+    instance = None
+
+    def __new__(cls, *args, **kwargs):
+        if not isinstance(cls.instance, cls):
```
@jujiang526 , I believe you've added this. Any thoughts?
